zio-http/src/main/scala/zio/http/QueryParams.scala

```diff
@@ -169,37 +159,111 @@ final case class QueryParams(map: Map[String, Chunk[String]]) {
   /**
    * Normalizes the query parameters by removing empty keys and values.
    */
-  def normalize: QueryParams =
-    if (isEmpty) self
-    else QueryParams(map.filter(i => i._1.nonEmpty && i._2.nonEmpty))
+  def normalize: QueryParams
 
   /**
    * Removes the specified key from the query parameters.
    */
-  def remove(key: String): QueryParams = QueryParams(map - key)
+  def remove(key: String): QueryParams
 
   /**
    * Removes the specified keys from the query parameters.
    */
-  def removeAll(keys: Iterable[String]): QueryParams = QueryParams(map -- keys)
+  def removeAll(keys: Iterable[String]): QueryParams
 
   /**
    * Converts the query parameters into a form.
    */
   def toForm: Form = Form.fromQueryParams(self)
+
 }
 
-object QueryParams {
+/**
+ * A ordered collection of query parameters.
+ */
+final case class ListMapQueryParams(map: ListMap[String, Chunk[String]]) extends QueryParams {
+  self =>
 
-  def apply(tuples: (String, Chunk[String])*): QueryParams =
-    QueryParams(map = Chunk.fromIterable(tuples).groupBy(_._1).map { case (key, values) =>
-      key -> values.flatMap(_._2)
+  /**
+   * Combines two collections of query parameters together. If there are
+   * duplicate keys, the values from both sides are preserved, in order from
+   * left-to-right.
+   */
+  override def ++(that: QueryParams): QueryParams =
```

Comment:
Would be good to add a test that `++` is order-preserving.

Reply:
Done.